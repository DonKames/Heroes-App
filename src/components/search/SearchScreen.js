import React, { useMemo } from 'react';
import { useNavigate, useSearchParams } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../heroes/HeroCard';

export const SearchScreen = () => {

    const [searchParams] = useSearchParams();

    const navigate = useNavigate();

    let q = '';

    if (searchParams.get('q') !== null) {
        q = searchParams.get('q');
    }

    const [values, handleInputChange] = useForm({
        searchHero: q
    });

    const { searchHero } = values;

    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

    const handleSearch = (e) => {
        e.preventDefault();
        navigate(`?q=${searchHero}`)
        console.log(searchHero);
    }

    return (
        <div>
            <h1>Search Screen</h1>
            <hr />
            <div className='row'>
                <div className='col-5'>
                    <h4>Search Form</h4>
                    <hr />
                    <form onSubmit={handleSearch}>
                        <input
                            autoComplete='off'
                            className='form-control'
                            name='searchHero'
                            onChange={handleInputChange}
                            placeholder='Find Your Hero'
                            type='text'
                            value={searchHero}
                        />
                        <div className='d-grid'>
                            <button
                                type='submit'
                                className='btn btn-outline-primary mt-2'
                            >
                                Search
                            </button>
                        </div>
                    </form>
                </div>
                <div className='col-7'>
                    <h4>Results</h4>
                    <hr />
                    {
                        (q === '')
                        &&
                        <div className='alert alert-info animate__animated animate__slideInUp'>
                            Search a hero
                        </div>
                    }

                    {
                        (q !== '' && heroesFiltered.length === 0)
                        &&
                        <div className='alert alert-danger animate__animated animate__tada'>
                            There is no heroes with: { q }
                        </div>
                    }

                    {
                        heroesFiltered.map(hero => (
                            <HeroCard
                                key={hero.id}
                                {...hero}
                            />
                        ))
                    }
                </div>
            </div>
        </div>
    );
};
