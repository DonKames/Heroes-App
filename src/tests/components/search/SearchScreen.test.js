import { act } from '@testing-library/react';
import { mount } from 'enzyme';
import { MemoryRouter, Route, Routes, useSearchParams } from 'react-router-dom';
import { AuthContext } from '../../../auth/AuthContext';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Pruebas <SearchScreen />', () => {


    test('Debe mostrarse correctamente', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Routes>
                    <Route path='/search' element={<SearchScreen />} />
                </Routes>
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();

        expect(wrapper.find('.alert-info').exists()).toBe(true);
    });

    test('Debe mostrar a Batman y el input con el valor del queryString.', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Routes>
                    <Route path='/search' element={<SearchScreen />} />
                </Routes>
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper.find('h5').text()).toBe('Batman');
        expect(wrapper.find('HeroCard').exists()).toBe(true)
        expect(wrapper).toMatchSnapshot();

    });

    test('Debe mostrar un error si no se encuentra el hero', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batmanFAKE']}>
                <Routes>
                    <Route path='/search' element={<SearchScreen />} />
                </Routes>
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-danger').exists()).toBe(true);
        expect(wrapper.find('.alert-danger').text().trim()).toBe('There is no heroes with: batmanFAKE');
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe llamar el handleSubmit', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batmanFAKE']}>
                <Routes>
                    <Route path='/search' element={<SearchScreen />} />
                </Routes>
            </MemoryRouter>
        );

        console.log(wrapper.html());

        
        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchHero',
                value: 'batman'
            }
        });
        
        
        act(() => {
            wrapper.find('form').prop('onSubmit')({
                preventDefault() { }
            })

        });

        expect(wrapper.render().find('h5').text()).toBe('Batman');
        expect(wrapper.find('input').prop('value')).toBe('batman');
    });
});
