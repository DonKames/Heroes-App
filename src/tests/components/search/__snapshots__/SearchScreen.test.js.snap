// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas <SearchScreen /> Debe mostrar a Batman y el input con el valor del queryString. 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchHero"
          onChange={[Function]}
          placeholder="Find Your Hero"
          type="text"
          value="batman"
        />
        <div
          className="d-grid"
        >
          <button
            className="btn btn-outline-primary mt-2"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col"
      >
        <div
          className="card animate__animated animate__fadeIn"
        >
          <div
            className="row"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="./assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas <SearchScreen /> Debe mostrar un error si no se encuentra el hero 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchHero"
          onChange={[Function]}
          placeholder="Find Your Hero"
          type="text"
          value="batmanFAKE"
        />
        <div
          className="d-grid"
        >
          <button
            className="btn btn-outline-primary mt-2"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger animate__animated animate__tada"
      >
        There is no heroes with: 
        batmanFAKE
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas <SearchScreen /> Debe mostrarse correctamente 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchHero"
          onChange={[Function]}
          placeholder="Find Your Hero"
          type="text"
          value=""
        />
        <div
          className="d-grid"
        >
          <button
            className="btn btn-outline-primary mt-2"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info animate__animated animate__slideInUp"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
